### ISSUE TITLE
False positive: used function is deemed “never used”

### ISSUE THREAD

**Issue by DoubleHyphen** [rust-lang/rust#142075]
### Code

```Rust
const _MY_CODE_IS_CORRECT: bool = true;

const fn sanity_check() {
    let true = _MY_CODE_IS_CORRECT else {
        panic!();
    };
}

const _SANITY_CHECK: () = sanity_check();
```

### Current output

```Shell
warning: function `sanity_check` is never used
 --> src/lib.rs:3:10
  |
3 | const fn sanity_check() {
  |          ^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Desired output

```Shell
Compiling with no warnings.
```

### Rationale and extra context

I have created some modules. I want each of them to fail to compile if it is not correct. To that end, I instituted a `sanity_check` in the crate, which is called within each module as `const _SANITY_CHECK: () = crate::sanity_check();`

The problem is that, because the `_SANITY_CHECK` consts are obviously not used anywhere, the entire `sanity_check` function is considered unused too. As one can tell, it is in fact very much used, and serves a very important purpose.

I guess this _could_ be worked around by renaming it to `_sanity_check`, but in this case leaving the function _actually_ unused would return no warnings, which is less than ideal.

### Other cases

```Rust

```

### Rust Version

```Shell
I tried it in the play-ground. I tried to find the version there but failed—that said, it's always the most recent one, isn't it?
```

**Comment by hanna-kruppe** [comment]
I would expect `const _: () = …;` to remove the warning but it doesn’t. Curiously, `pub const _` doesn’t get the warning even though there’s no name to make public so the `pub` shouldn’t make a difference. That definitely feels like a bug to me.

**Comment by fmease** [comment]
To recap this reduces to

```rs
const fn f() {} //~ WARN function `f` is never used
const _X: () = f();
```

However, we apparently don't do this for any other cases either (namely, propagating liveness via `_ident`):

```rs
fn f() {} //~ WARN function `f` is never used
fn _g() { f() }
```

**Comment by fmease** [comment]
Also CC https://github.com/rust-lang/rust/issues/118424

**Comment by hanna-kruppe** [comment]
I think it’s a consistent, defensible choice to treat an underscore prefix as only suppressing the dead code lint for that one item, without counting as a “root” for the purpose of computing which things are reachable from exports/entry points. It could be done differently but there’s arguments for it as well.

However, `const _` not counting as “root” seems wrong because that syntax exists *only* for the side effects of type checking and evaluating the constant expression. It’s the best solution for “I only want a compile time assertion, not define a value” (at the module level; within functions a const block is more flexible) and was [specifically added after `const _FOO` already existed](https://rust-lang.github.io/rfcs/2526-const-wildcard.html). So it should work for @DoubleHyphen’s use case and for https://github.com/rust-lang/rust/issues/118424 — even and especially if we stick with the current behavior for `const _FOO` for consistency with other items.

**Comment by DoubleHyphen** [comment]
> Curiously, `pub const _` doesn’t get the warning

…it does when it's inside a module, though!

```rust
const fn foo() {} // warning: function `foo` is never used

mod whatever {
    pub const _: () = crate::foo();
}
```

**Comment by Jules-Bertholet** [comment]
> `pub const _` doesn’t get the warning even though there’s no name to make public so the `pub` shouldn’t make a difference.


Wait, why is that (visibility modifier on an unnamed `const` item) even allowed? Should we perhaps deprecate it once this issue is fixed? (It seems that there are so few uses, we maybe could even get away with a full removal: https://grep.app/search?q=pub+const+_%3A)

**Comment by peter-lyons-kehl** [comment]
@Jules-Bertholet et al, `pub const _` is planned to give a warning. On current `nightly` (`rustc 1.97.0-nightly (e8e4541ff 2026-04-15)`:
```
warning: visibility qualifiers have no effect on `const _` declarations
  --> src/lib.rs:45:1
   |
45 | pub const _: () = {};
   | ^^^ help: remove the qualifier
```

**Comment by peter-lyons-kehl** [comment]
Workaround (if your MSRV is recent, see below, and if you don't want to use `#[allow(dead_code)]` or you want/have to have `#![forbid(dead_code)])`: *assign* such an item to a `_` constant. Or use it in an expression/block that is then stored in a `_` constant. For example:
```rust
const _: fn(&'static str)->bool = is_exact_version;

const fn is_exact_version(expected_version: &'static str) -> bool {
    // We can't use a comparison operator ==, because as of April 2026 trait PartialEq is not const.
    matches!(expected_version.as_bytes(), b"0.1.0")
}
```
Or, without repeating the type, assign to a variable:
```rust
const _: () = {
    let _ = is_exact_version;
    let _ = another_victim_here;
    // ...
}
```

Both ways work on current stable 1.95.0, but NOT on 1.88.0.

**Comment by Jules-Bertholet** [comment]
> @Jules-Bertholet et al, `pub const _` is planned to give a warning. 

I'm aware, I implemented it 😆 https://github.com/rust-lang/rust/pull/147136

**Comment by Jules-Bertholet** [comment]
Given that `const _` now removes the warning (https://github.com/rust-lang/rust/pull/142208), I think this can be closed?

**Comment by peter-lyons-kehl** [comment]
Thank you. OK with me to close.
